### PR TITLE
[performance] faster Tree expand - eclipse.platform.swt#901

### PR DESCRIPTION
**Pull request by jukzi**
- insert every item at index 0 (in reverse order)

https://github.com/eclipse-platform/eclipse.platform.swt/issues/901

### CODE REVIEW

**Comment by laeubi** [comment]
@SyntevoAlex should this happen for all platforms or only for Windows?

**Comment by akurtakov** [comment]
Jface should be ws agnostic.

**Comment by laeubi** [comment]
> Jface should be ws agnostic.

That's true but still we can have platform switches in it if required.

**Comment by SyntevoAlex** [comment]
I suggest adding a code comment with explanation.

> should this happen for all platforms or only for Windows?

Doing it on all platforms won't hurt, I think. But I didn't test it myself. You could play with my `Bug575787_Tree_SetItemCount_Perf` snippet (already in SWT)

**Comment by jukzi** [comment]
> I suggest adding a code comment with explanation.

i added comments and reverse order

**Comment by SyntevoAlex** [comment]
Good. Note that I didn't test anything.

**Review by SyntevoAlex — COMMENTED** [review]
Looks reasonable to me, but I didn't test it.

**Comment by jukzi** [comment]
I also added the setRedraw(false) trick, which cuts downs setImage() and setText() time when i sort the Problems view with many elements. for example ~4000 elements org.eclipse.swt.widgets.Display.readAndDispatch () reduces from 1300ms to 600ms

File: bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/TreeViewer.java [hunk 1]
```diff
@@ -363,7 +363,9 @@ void preservingSelection(Runnable updateCode, boolean reveal) {
 		// updateCode.run():
 		insidePreservingSelection = true;
 		try {
+			tree.setRedraw(false);
 			super.preservingSelection(updateCode, reveal);
+			tree.setRedraw(true);
```
Comment (basilevs) [inline bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/TreeViewer.java]: Should `setRedraw(true)` be in a finally block?

**Comment by basilevs** [comment]
I've tried the test reproducing the problem I'm working on from https://github.com/eclipse-platform/eclipse.platform.ui/issues/649
## GTK
No regressions
### Before change
﻿210 ms to process 10000 items
202 ms to process 100000 items
The time to select scales as O(n^-0.02) as the size of the model grows from 10000 to 100000
1160 ms to process 10000 items
83267 ms to process 100000 items
The time to reveal scales as O(n^1.86) as the size of the model grows from 10000 to 100000
### After change
196 ms to process 10000 items
399 ms to process 100000 items
The time to select scales as O(n^0.31) as the size of the model grows from 10000 to 100000
1260 ms to process 10000 items
86350 ms to process 100000 items
The time to reveal scales as O(n^1.84) as the size of the model grows from 10000 to 100000
## MacOS
Slowdown?
### Before the change
53 ms to process 10000 items
430 ms to process 100000 items
The time to select scales as O(n^0.90) as the size of the model grows from 10000 to 100000
38 ms to process 10000 items
381 ms to process 100000 items
The time to reveal scales as O(n^1.00) as the size of the model grows from 10000 to 100000
### After the change
110 ms to process 10000 items
1005 ms to process 100000 items
The time to select scales as O(n^0.96) as the size of the model grows from 10000 to 100000
91 ms to process 10000 items
944 ms to process 100000 items
The time to reveal scales as O(n^1.01) as the size of the model grows from 10000 to 100000

**Comment by basilevs** [comment]
### MacOS increased tree size
The change definitely increases a time to reveal an item in a large tree
### Before
413 ms to process 100000 items
4265 ms to process 1000000 items
The time to select scales as O(n^1.01) as the size of the model grows from 100000 to 1000000
389 ms to process 100000 items
4133 ms to process 1000000 items
The time to reveal scales as O(n^1.03) as the size of the model grows from 100000 to 1000000
### After
969 ms to process 100000 items
9748 ms to process 1000000 items
The time to select scales as O(n^1.00) as the size of the model grows from 100000 to 1000000
963 ms to process 100000 items
9959 ms to process 1000000 items
The time to reveal scales as O(n^1.01) as the size of the model grows from 100000 to 1000000

**Comment by jukzi** [comment]
> The change definitely increases a time to reveal an item in a large tree

@SyntevoAlex any idea?
@basilevs can you please share the testprogram? i want to try it on windows

**Comment by basilevs** [comment]
@juksi https://github.com/eclipse-platform/eclipse.platform.ui/issues/649

https://github.com/eclipse-platform/eclipse.platform.ui/files/11019153/bug516281.zip

**Comment by jukzi** [comment]
> @Juksi #649

jukZi :-)
thanks,. can you post measurements if you combine this PR with 649? all platforms, before+after 
This PR contains 3 different optimizations: 

- move setExpanded() after expanding the children
- insert every item at index 0 (in reverse order)
- setRedraw

can you please figure out which of them does cause regression on Mac?

**Comment by basilevs** [comment]
> can you post measurements if you combine this PR with 649?

#649 is WIP, there is nothing to combine, I've tested on master and on your branch
> all platforms, before+after

I don't have Windows env and have already covered Linux and MacOS

> This PR contains 3 different optimizations: 
>     * move setExpanded() after expanding the children
>     * insert every item at index 0 (in reverse order)
>     * setRedraw
> can you please figure out which of them does cause regression on Mac?

I've removed new `setRedraw()` lines from `TreeViewer` and got old performance:

```
395 ms to process 100000 items
4172 ms to process 1000000 items
The time to select scales as O(n^1.02) as the size of the model grows from 100000 to 1000000
366 ms to process 100000 items
4070 ms to process 1000000 items
The time to reveal scales as O(n^1.05) as the size of the model grows from 100000 to 1000000
```

This is a major surprise - I assumed `setRedraw()` was a natural wrapper for any batch operations.

**Comment by basilevs** [comment]
Performance tests do not work, but are necessary. Life is pain.

**Comment by jukzi** [comment]
on windows:
```
before PR:
3 ms to process 10000 items
6 ms to process 100000 items
The time to select scales as O(n^0,24) as the size of the model grows from 10000 to 100000
1094 ms to process 10000 items
10413 ms to process 100000 items
The time to reveal scales as O(n^0,98) as the size of the model grows from 10000 to 100000

without redraw(false)
3 ms to process 10000 items
6 ms to process 100000 items
The time to select scales as O(n^0,27) as the size of the model grows from 10000 to 100000
995 ms to process 10000 items
9993 ms to process 100000 items
The time to reveal scales as O(n^1,00) as the size of the model grows from 10000 to 100000

with redraw(false)
5 ms to process 10000 items
13 ms to process 100000 items
The time to select scales as O(n^0,38) as the size of the model grows from 10000 to 100000
3105 ms to process 10000 items
30691 ms to process 100000 items
The time to reveal scales as O(n^0,99) as the size of the model grows from 10000 to 100000
```

=> that test can not show that this PR  can make some things faster, but also no regression
=> with this test redraw(false) is not good. probably because the tested view is much less complicated then the Problem View i measured with. So probably it's better to use redraw(false) not in general but may be in Problem View

**Comment by basilevs** [comment]
@jukzi  `reveal()` operation is not used in Problem View. Project Explorer or Outline views are a better match for my test.

**Comment by jukzi** [comment]
let forget about redraw(false), it's already set in org.eclipse.ui.internal.views.markers.UIUpdateJob.runInUIThread(IProgressMonitor)

**Review by basilevs — APPROVED** [review]

**Comment by basilevs** [comment]
@jukzi have you tried using` TreeItem.setItemCount()` instead of creating items one-by-one? IMO, it should dramatically increase performance.

**Comment by jukzi** [comment]
> @jukzi have you tried using` TreeItem.setItemCount()` instead of creating items one-by-one? IMO, it should dramatically increase performance.

I have not tried that. i don't understand how it should be used. Is that replacement for multiple `new TreeItem(parent, -1)` ?

**Comment by basilevs** [comment]
> Is that replacement for multiple `new TreeItem(parent, -1)` ?

Yes. The change would be to replace multiple new `TreeItem()` with `TreeItem.setItemCount()`
than traverse those with `getItems()`, `updateItem()`and `updatePlus()`. Performance gain is not guaranteed and would require testing (as `getItems()` may be slow enough to negate any gains).

**Comment by jukzi** [comment]
We got a Problem: this change keeps failing on linux - and only linux  - 

[org.eclipse.jface.tests.viewers.VirtualLazyTreeViewerTest.testExpandToLevel](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-1331/8/testReport/junit/org.eclipse.jface.tests.viewers/VirtualLazyTreeViewerTest/testExpandToLevel/)
[org.eclipse.jface.tests.viewers.TreeViewerWithLimitTest.testRemoveItemsAtParent](https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-1331/8/testReport/junit/org.eclipse.jface.tests.viewers/TreeViewerWithLimitTest/testRemoveItemsAtParent/)

The code should be theoretically OS independent. i am on windows. @basilevs can you look why it fails on linux?

**Comment by laeubi** [comment]
@jukzi first basic step would be to enhance the test with *what* assertions failed and probably adding a message that helps in understand the test failure.

**Comment by basilevs** [comment]
The test fails to expand second level of the tree:
Screenshot 2023-12-01 at 13 55 19

**Comment by jukzi** [comment]
> Yes. The change would be to replace multiple new `TreeItem()` with `TreeItem.setItemCount()`
> than traverse those with `getItems()`, `updateItem()`and `updatePlus()`. Performance gain is not guaranteed and would require testing (as `getItems()` may be slow enough to negate any gains).

that seems to be not usable in TreeViewer  because
org.eclipse.jface.viewers.TreeViewer.newItem(Widget, int, int) is protected - i.e. in principle it could have been overloaded by clients.
Even though `@noextend This class is not intended to be subclassed by clients.` the reality looks different:
image

**Comment by jukzi** [comment]
linux VirtualLazyTreeViewerTest.testExpandToLevel
> The test fails to expand second level of the tree:

but why - @SyntevoAlex ?

**Comment by basilevs** [comment]
`newItem()` is already not used in `org.eclipse.jface.viewers.TreeViewer.setChildCount(Object, int)` used by lazy TreeViewer and likely in other places.

**Comment by basilevs** [comment]
None of new `createTreeItem()` lines are called in the test. So problem is probably with move of `setExpanded()`

UPDATE: confirmed, moving it to the old spot fixes one test (I only check one test)

**Comment by jukzi** [comment]
> UPDATE: confirmed, moving it to the old spot fixes one test (I only check one test)

i will split that commit

**Comment by basilevs** [comment]
There is a hint in Tree:
https://github.com/basilevs/eclipse.platform.swt/blob/4f69de6866723e1e139a5762d993c5dabf2c4799/bundles/org.eclipse.swt/Eclipse%20SWT/gtk/org/eclipse/swt/widgets/Tree.java#L2545-L2566
```java
long gtk_row_has_child_toggled (long model, long path, long iter) {
	/*
	* Feature in GTK. The expanded state of a row that lost
	* its children is not persisted by GTK. So, the row
	* doesn't exhibit the expanded state after obtaining the
	* children. The fix is to preserve the expanded state
	* and use this callback, as it is invoked when a row has
	* gotten the first child row or lost its last child row.
	*/
	int [] index = new int [1];
	GTK.gtk_tree_model_get (modelHandle, iter, ID_COLUMN, index, -1);
	if (index [0] >= items.length) return 0;
	TreeItem item = items [index [0]];
	if (item == null) return 0;
	int childCount = GTK.gtk_tree_model_iter_n_children (modelHandle, item.handle);
	if (childCount != 0 && item.isExpanded) {
		OS.g_signal_handlers_block_matched (handle, OS.G_SIGNAL_MATCH_DATA, 0, 0, 0, 0, TEST_EXPAND_ROW);
		GTK.gtk_tree_view_expand_row (handle, path, false);
		OS.g_signal_handlers_unblock_matched (handle, OS.G_SIGNAL_MATCH_DATA, 0, 0, 0, 0, TEST_EXPAND_ROW);
	}
	return 0;
}
```
But the condition was not used during my debugging.

**Comment by basilevs** [comment]
Idea - Tree does not recheck states of children when item is expanded. So if state did not stick to GTK model the first time, when its parent is expanded the expansion state is not reapplied.

Effectively the tree only works as expected when it is manipulated in a way end user would - top to bottom.
This seems to be a bug in GTK Tree implementation, as other platforms persist expanded state of invisible item.

**Comment by jukzi** [comment]
> None of new `createTreeItem()` lines are called in the test. So problem is probably with move of `setExpanded()`

i created an experiment: https://github.com/eclipse-platform/eclipse.platform.ui/pull/1357/files
it seems to be SLOWER on windows to use setItemCount

**Comment by jukzi** [comment]
> This seems to be a bug in GTK Tree implementation, as other platforms persist expanded state of invisible item.

sounds reasonable

**Comment by jukzi** [comment]
only unrelated test errors: https://github.com/eclipse-platform/eclipse.platform.ui/issues/1351

**Comment by laeubi** [comment]
> > This seems to be a bug in GTK Tree implementation, as other platforms persist expanded state of invisible item.
> 
> sounds reasonable

Why is it not a bug in **windows** that expanding invisible childs is faster than expand visible ones? Just a thought ;-)

**Comment by jukzi** [comment]
> Why is it not a bug in **windows** that expanding invisible childs is faster than expand visible ones? Just a thought ;-)

Invisible things don't need to be drawn.

**Comment by SyntevoAlex** [comment]
> linux VirtualLazyTreeViewerTest.testExpandToLevel
> 
> > The test fails to expand second level of the tree:
> 
> but why - @SyntevoAlex ?

I glanced at `gtk_tree_view_real_expand_row()` in GTK sources and I had the impression that "item is expanded" and "internal data for child items exist" are the same thing. In other words, child items are only created when parent is expanded. This might explain why it's not possible to modify item that is not expanded yet.

If you care to make a pure SWT snippet, I could spend a bit of time to check it with native debugger.

**Comment by jukzi** [comment]
> If you care to make a pure SWT snippet, I could spend a bit of time to check it with native debugger.

@basilevs can you create a pure SWT reproducer for GTK?

**Comment by basilevs** [comment]
@jukzi no, but I assume https://github.com/basilevs/eclipse.platform.swt/blob/158e4c6d7e70a9c305ea3e788928b7ff6419edb6/tests/org.eclipse.swt.tests/JUnit%20Tests/org/eclipse/swt/tests/junit/performance/Test_org_eclipse_swt_widgets_Tree.java#L262 would work fine. Try it to see if results match your report.

**Comment by basilevs** [comment]
Following test fails on last line on GTK. If user manually "helps" to expand "item2" node, the test instantly passes.
It does not fail if order of expansions is reversed. 
```diff
diff --git a/tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Tree.java b/tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Tree.java
index 0ad17d5..86ada48 100644
--- a/tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Tree.java
+++ b/tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_widgets_Tree.java
@@ -1188,6 +1188,35 @@
 		new TreeItem(item_0, 0, 0);
 		assertEquals(10, item_0.getItemCount());
 	});
 }
 
+@Test
+public void test_persistExpandStatusForInvisibleItems() throws InterruptedException {
+	tree.dispose();
+	tree = new Tree(shell, SWT.VIRTUAL);
+	setWidget(tree);
+	shell.setLayout(new FillLayout());
+	SwtTestUtil.openShell(shell);
+	int itemCount[] = new int[] {1};
+	List<TreeItem> dataRequests = new ArrayList<>();
+	tree.addListener(SWT.SetData, event -> {
+		TreeItem item = (TreeItem) event.item;
+		dataRequests.add(item);
+		item.setText("item"+itemCount[0]++);
+		item.setItemCount(1);
+	});
+	TreeItem item1 = new TreeItem(tree, SWT.NONE);
+	TreeItem item2 = item1.getItem(0);
+	TreeItem item3 = item2.getItem(0);
+	assertTrue(dataRequests.remove(item1));
+	assertTrue(dataRequests.remove(item2));
+	assertTrue(dataRequests.isEmpty());
+	item2.setExpanded(true);
+	assertTrue(dataRequests.isEmpty());
+	item1.setExpanded(true);
+	// Item 3 is now visible and should request data
+	SwtTestUtil.processEvents(10000, () -> !dataRequests.isEmpty());
+	assertFalse(dataRequests.isEmpty());
+}
+
 }
```

**Comment by jukzi** [comment]
ignoring random fail https://github.com/eclipse-platform/eclipse.platform.ui/issues/275

**Pull request merged**